cloud-aws-common/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/common/connector/resource/AwsResourceVolumeConnector.java

```diff
@@ -0,0 +1,72 @@
+import software.amazon.awssdk.services.ec2.model.DeleteVolumeRequest;
+import software.amazon.awssdk.services.ec2.model.DescribeVolumesRequest;
+import software.amazon.awssdk.services.ec2.model.DescribeVolumesResponse;
+import software.amazon.awssdk.services.ec2.model.DetachVolumeRequest;
+
+@Service
+public class AwsResourceVolumeConnector implements ResourceVolumeConnector {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(AwsResourceVolumeConnector.class);
+
+    @Override
+    public void detachVolumes(AuthenticatedContext authenticatedContext, List<CloudResource> cloudResources) throws Exception {
+        Map<String, List<Volume>> instanceVolumeIdsMap = cloudResources.stream().filter(res -> null != res.getInstanceId())
+                .collect(toMap(CloudResource::getInstanceId, resource -> resource.getParameter(CloudResource.ATTRIBUTES,
+                        VolumeSetAttributes.class).getVolumes()));
+        AmazonEc2Client amazonEC2Client = getEc2Client(authenticatedContext);
+        for (Entry<String, List<Volume>> entry: instanceVolumeIdsMap.entrySet()) {
+            entry.getValue().forEach(volume -> {
+                DetachVolumeRequest detachVolumeRequest = DetachVolumeRequest.builder().instanceId(entry.getKey())
+                        .volumeId(volume.getId()).device(volume.getDevice()).build();
+                amazonEC2Client.detachVolume(detachVolumeRequest);
+                boolean available = false;
+                while (!available) {
+                    DescribeVolumesRequest describeVolumesRequest = DescribeVolumesRequest.builder().volumeIds(List.of(volume.getId())).build();
+                    DescribeVolumesResponse response = amazonEC2Client.describeVolumes(describeVolumesRequest);
+                    if (response.hasVolumes()) {
+                        software.amazon.awssdk.services.ec2.model.Volume volumeResponse = response.volumes().get(0);
+                        available = "available".equals(volumeResponse.state().toString());
+                    }
+                }
+            });
+        }
+    }
+
+    @Override
+    public void deleteVolumes(AuthenticatedContext authenticatedContext, List<CloudResource> cloudResources) throws Exception {
+        Map<String, List<Volume>> instanceVolumeIdsMap = cloudResources.stream().filter(res -> null != res.getInstanceId())
+                .collect(toMap(CloudResource::getInstanceId, resource -> resource.getParameter(CloudResource.ATTRIBUTES,
+                        VolumeSetAttributes.class).getVolumes()));
+        AmazonEc2Client amazonEC2Client = getEc2Client(authenticatedContext);
+        for (Entry<String, List<Volume>> entry: instanceVolumeIdsMap.entrySet()) {
+            entry.getValue().forEach(volume -> {
+                DeleteVolumeRequest deleteVolumeRequest = DeleteVolumeRequest.builder().volumeId(volume.getId()).build();
+                amazonEC2Client.deleteVolume(deleteVolumeRequest);
+            });
+        }
```

Comment:
Are we assuming that delete operation is idempotent because This block can be partially successful which means few volumes for few instances can be deleted and then we get a failure. In that case, it will only throw exception without giving any information on partial success.